pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -529,38 +530,51 @@ func (s *Controller) WorkloadInstanceHandler(wi *model.WorkloadInstance, event m
 	fullPush := false
 	for _, cfg := range cfgs {
 		se := cfg.Spec.(*networking.ServiceEntry)
-		if se.WorkloadSelector == nil || !labels.Instance(se.WorkloadSelector.Labels).SubsetOf(wi.Endpoint.Labels) {
-			// Not a match, skip this one
-			continue
-		}
-		seNamespacedName := config.NamespacedName(cfg)
-		services := s.services.getServices(seNamespacedName)
-		instance := convertWorkloadInstanceToServiceInstance(wi, services, se)
-		instances = append(instances, instance...)
-		if addressToDelete != "" {
-			for _, i := range instance {
-				di := i.DeepCopy()
-				di.Endpoint.Address = addressToDelete
-				instancesDeleted = append(instancesDeleted, di)
-			}
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else if event == model.EventDelete {
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else {
-			s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, instance)
-		}
-		// If serviceentry's resolution is DNS, make a full push
-		// TODO: maybe cds?
-		if (se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN) &&
-			se.WorkloadSelector != nil {
-
-			fullPush = true
-			for _, inst := range instance {
-				configsUpdated[model.ConfigKey{
-					Kind:      kind.ServiceEntry,
-					Name:      string(inst.Service.Hostname),
-					Namespace: cfg.Namespace,
-				}] = struct{}{}
+		if se.WorkloadSelector != nil {
+			// If the workload selector exists, then we need to verify that the subset still is a match (in case the workload instance labels change)
+			if reflect.DeepEqual(se.WorkloadSelector.Labels, wi.Endpoint.Labels) || labels.Instance(se.WorkloadSelector.Labels).SubsetOf(wi.Endpoint.Labels) {
```

Comment:
`reflect.DeepEqual(se.WorkloadSelector.Labels, wi.Endpoint.Labels) `

This is not needed, and not I meant.